Find zip file end of central directory backwards up to max possible size

carlossanlop opened the pull request:
Fixes https://github.com/dotnet/corefx/issues/34892 and https://github.com/dotnet/corefx/issues/37394

### Issue

Huge files without an end of central directory take a long time to throw because we seek to the end then search backwards for the EOCD signature all the way to the end.

### Fix

I addressed the problem by ensuring we only search for the signature within the maximum possible number of bytes where it can be found.

I applied the same principle to the Zip64 End of Central Directory Locator search, which happened to do be doing a similar search.

I also moved the code that tries to find the Zip64 EOCD to its own function, to make the code look cleaner.

### Pending

- The file for the unit test I currently tested is huge, so I have to upload it to corefx-testdata first, but when I ran it locally, the 549 MB file provided in #34892 took ~100 milliseconds to throw after not finding the EOCD signature.
- Add at least two more zip files, or make sure we already have UTs that test:
    - A huge, malformed zip file with bytes truncated at the end.
    - An empty zip file.


CC @riverar @kburgoyne

ericstj commented on src/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs:
```diff
@@ -9,6 +9,22 @@ namespace System.IO.Compression.Tests
+
+        //    Assert.Throws<InvalidDataException>(() =>
+        //    {
+        //        using FileStream fs = File.OpenRead(@"large.cab"); // 549 MB
```
Do you really need actual files to test this, or can you just construct streams in memory?

carlossanlop replied:
Let me see if I can create a huge file without an EOCD. If it's too much, I would prefer to save the file instead.

ericstj replied:
Let me be more clear about this: please don't add a 500 MB file to our test data for a single test.  Cost/benefit is not worth it.  Test data is downloaded by all our contributors and not everyone has broadband.  Test data should be used sparingly.

jonwhite1 replied:
I don't know if it helps but you could make a fake Stream class that claims to be 1MB long, then unzip it, and make sure it doesn't call Read more than about 64K / n times, where n is the size of the reads.  My VSO bug has a stream implementation in the repro that just counts reads and seeks if you want a starting point.

ericstj commented on src/System.IO.Compression/tests/ZipArchive/zip_ReadTests.cs:
```diff
@@ -9,6 +9,22 @@ namespace System.IO.Compression.Tests
+        //[Fact]
+        //public static void ReadHugeFileWithoutEOCD()
+        //{
+        //    Diagnostics.Stopwatch sw = new Diagnostics.Stopwatch();
```
We cannot test with stopwatch during CI.  No matter how much we relax this it would still be flaky.  Instead have a test stream that reports how much data actually was read out of it, then assert that less than some amount was read.

carlossanlop replied:
Thanks for letting me know, I understand, I wasn't planning on leaving this. I was just testing the elapsed time to compare with the original reports (tens of seconds, or even minutes).

riverar commented:
⚠ This change breaks compatibility with executables, as mentioned in #34892, where the signature could exist near the end of the file.

jonwhite1 commented:
Hi Riverar - if an executable had an embedded zip file as a resource, even though this might find the EOCD record, wouldn't extraction still fail because the pointer to the start of the central directory (and all the other pointers) would be wrong?

riverar commented:
@jonwhite1 Sounds right, however enumerating entries / use of those pointers is not required. An app, that instantiates `ZipArchive` around an executable as part of some scheme to determine if an archive exists, would suddenly throw with this fix in place. It's contrived, I know, just flagging it for consideration. (Am okay with writing off that scenario as unsupported too.)

carlossanlop commented on src/System.IO.Compression/src/System/IO/Compression/ZipArchive.cs:
```diff
@@ -530,9 +530,15 @@ private void ReadEndOfCentralDirectory()
+                // But if the EOCD has max possible size, the signature should be found somewhere in the previous 64K + 4 bytes
+                if (!ZipHelper.SeekBackwardsToSignature(_archiveStream,
+                        ZipEndOfCentralDirectoryBlock.SignatureConstant,
+                        ZipEndOfCentralDirectoryBlock.ZipFileCommentMaxLength + ZipEndOfCentralDirectoryBlock.SignatureSize))
```
@ericstj as we discussed earlier in person, I double checked this restriction.

The maximum size of the EOCD zip file comment section is not explicitly mentioned in the [Zip File Format Specification](https://pkware.cachefly.net/webdocs/casestudies/APPNOTE.TXT):

```
4.4.25 .ZIP file comment length: (2 bytes)
      The length of the comment for this .ZIP file.

4.4.26 .ZIP file comment: (Variable)
      The comment for this .ZIP file.  ZIP file comment data
      is stored unsecured.  No encryption or data authentication
      is applied to this area at this time.  Confidential information
      SHOULD NOT be stored in this section.
```

But I think we should make the assumption that the maximum allowed length is 64K for the following reasons:

1. The File Header (4.3.12) has two similar fields: "_File Comment Length (2 bytes)_" and "_File Comment (variable size)_" and it has an explicitly described size restriction combined with two other variable length fields:

```
4.4.10 file name length: (2 bytes)
4.4.11 extra field length: (2 bytes)
4.4.12 file comment length: (2 bytes)
       The length of the file name, extra field, and comment
       fields respectively.  The combined length of any
       directory record and these three fields SHOULD NOT
       generally exceed 65,535 bytes.  If input came from standard
       input, the file name length is set to zero.  

4.4.18 file comment: (Variable)
       The comment for this file.
```


2. [WinZip restricts](http://kb.winzip.com/help/winzip/help_comment.htm) the Zip File Comments to 64,000 characters:

> _Zip file comments are limited in length. If you have recent versions of the Windows "Rich Edit" control installed, the limit is about 64,000 characters; otherwise, the limit is about 48,000 characters. For greatest compatibility with other users' systems and with other Zip file utilities, we recommend limiting comments to 48,000 characters._


3. [Java (OpenJDK 8u60)](http://hg.openjdk.java.net/jdk8u/jdk8u60/jdk/file/935758609767/src/share/classes/java/util/zip/ZipOutputStream.java#l140) restricts it to 64K:

```java
/**
 * Sets the ZIP file comment.
 * @param comment the comment string
 * @exception IllegalArgumentException if the length of the specified
 *            ZIP file comment is greater than 0xFFFF bytes
 */
public void setComment(String comment)
{
    if (comment != null)
    {
        this.comment = zc.getBytes(comment);
        if (this.comment.length > 0xffff)
            throw new IllegalArgumentException("ZIP file comment too long.");
    }
}
```

4. [Python](https://docs.python.org/3/library/zipfile.html#zipfile.ZipFile.comment) restricts it to 64K:
> _ZipFile.comment_
>     _The comment associated with the ZIP file as a bytes object. If assigning a comment to a ZipFile instance created with mode 'w', 'x' or 'a', it should be no longer than 65535 bytes. Comments longer than this will be truncated._

carlossanlop commented:
@riverar @ericstj @jonwhite1 

I did some manual tests and compared them to the behavior of .NET Core (both with and without the fix).
Below is each file I used (describing their contents) and the results on different compression tools and language libraries:

**SelfExtracting.exe** - _created with WinZip_
>Normal.zip (1KB)
>>file.txt (1KB)

    OK?     - 7-Zip - Opens and extracts successfully, but shows message: "Warning: There are some data after the end of the payload data.".
    OK      - WinRAR - Opens and extracts successfully.
    OK      - WinZip - Opens and extracts successfully.
    N/A     - Windows Explorer - Can't open an EXE as a Zip.
    OK      - unzip (Ubuntu) - Opens and extracts successfully.
    OK      - Python 3.7.4 - Opens and extracts successfully.
    FAIL    - Java OpenJDK 12.0.2 (2019-07-16) - Can open the exe as a zip file but can't find anything inside (invalid zip structure).
    OK      - corefx (without fix) - It eventually finds a zip file inside.
    OK      - corefx (with fix) - It eventually finds a zip file inside, but only because the new limits are not reached (small file).


**SelfExtracting.exe** - _created with WinZip_
>EOCDCommentOf64KB.zip (68KB) - _comment added via Python, (max length is 65,535 ASCII chars (64KB))_
>>file.txt (1KB)

    OK?     - 7-Zip - Open and extracts successfully, but shows message: "Warning: There are some data after the end of the payload data.".
    OK      - WinRAR - Opens and extracts successfully.
    OK      - WinZip - Opens and extracts successfully.
    N/A     - Windows Explorer - Can't open an EXE as a Zip.
    OK      - unzip (Ubuntu) - Opens and extracts successfully.
    FAIL    - Python 3.7.4 - Can open the exe as a zip file but can't find anything inside (invalid zip structure).
    FAIL    - Java OpenJDK 12.0.2 (2019-07-16) - Can open the exe as a zip file but can't find anything inside (invalid zip structure).
    FAIL    - .NET Core 3.0 - Can open the file but can't find anything inside (invalid zip structure).
    FAIL    - corefx (with fix) - Can open the file but can't find anything inside (invalid zip structure).

######  _Note: I found [a StackOverflow thread](https://stackoverflow.com/a/7955605/8816314) where they explained that the workaround for Java is rather complicated, and probably the same workaround applies to C# and Python: Basically, manually ignore exe file sections so the file can be treated as zip.



**EOCDCommentOf32KB.zip** - _comment added via WinRAR (max length is 32KB)_
>file.zip (1KB)

    OK      - 7-Zip - Opens and extracts successfully.
    OK      - WinRAR - Opens and extracts successfully.
    OK      - WinZip  - Opens and extracts successfully.
    OK      - Windows Explorer - Opens and extracts successfully.
    OK      - unzip (Ubuntu) - Opens and extracts successfully.
    OK      - Python 3.7.4 - Opens and extracts successfully.
    OK      - Java OpenJDK 12.0.2 (2019-07-16) - Opens and extracts successfully.
    OK      - corefx (without fix) - Opens and extracts successfully.
    OK      - corefx (with fix) - Opens and extracts successfully.

**EOCDCommentOf65KChars.zip** - _comment added via WinZip, (max length is 65,000 ASCII chars)_
>file.zip (1KB)

    OK      - 7-Zip - Opens and extracts successfully.
    OK      - WinRAR - Opens and extracts successfully.
    OK      - WinZip - Opens and extracts successfully.
    OK      - Windows Explorer - Opens and extracts successfully.
    OK      - unzip (Ubuntu) - Opens and extracts successfully.
    OK      - Python 3.7.4 - Opens and extracts successfully.
    OK      - Java OpenJDK 12.0.2 (2019-07-16) - Opens and extracts successfully.
    OK      - corefx (without fix) - Opens and extracts successfully.
    OK      - corefx (with fix) - Opens and extracts successfully.


**EOCDCommentOf64KB.zip** - _comment added via Python, (max length is 65,535 ASCII chars (64KB))_
>file.zip (1KB)

    OK      - 7-Zip - Fast to open and extract. Cannot show comment to user.
    OK?     - WinRAR - Fast to open and extract. Silently fails (shows empty string) when attempting to read comment (beyond their defined limit).
    OK?     - WinZip - Fast to open and extract. Shows failure window when attempting to read comment (too long for them).
    OK?     - Windows Explorer - Slow to open and extract. Cannot show comment to user.
    OK      - unzip (Ubuntu) - Fast to open and extract. Prints comment automatically.
    OK      - Python 3.7.4 - Fast to open and extract. Has public API to access comment string.
    OK      - Java OpenJDK 12.0.2 (2019-07-16) - Fast to open and extract. Has public API to access comment string.
    OK?     - corefx (without fix) - Slow to open and extract. Does not have public API to access comment string.
    OK      - corefx (with fix) - Fast to open and extract: BUG FIXED!. Does not have public API to access comment string.

**large.cab** (549 MB) - _downloaded from original issue (cabs are not a zip file, so they don't have an EOCD)_
>_...Various files and folders..._

    OK      - 7-Zip - Fast to open and extract (knows how to open cabs).
    OK      - WinRAR - Fast to open and extract (knows how to open cabs).
    OK      - WinZip - Moderately fast to open and extract (knows how to open cabs).
    OK      - Windows Explorer - Fast to open and extract (knows how to open cabs).
    FAIL    - unzip (Ubuntu) - Fast to detect invalid zip structure.
    FAIL    - Python 3.7.4 - Fast to detect invalid zip structure.
    FAIL    - Java OpenJDK 12.0.2 (2019-07-16) - Fast to detect invalid zip structure.
    FAIL    - corefx (without fix) - Takes 3 minutes to detect invalid zip structure.
    FAIL    - corefx (with fix) - Fast to detect invalid zip structure: BUG FIXED!


**LargeWithEOCDCommentOf64KB.zip** - _comment added via Python, (max length is 65,535 ASCII chars (64KB))_
>large.cab (549 MB) - _downloaded from original issue (cabs are not a zip file, so they don't have an EOCD)_
>>_...Various files and folders..._

    OK    - 7-Zip - Fast to open and extract. Cannot show comment to user.
    OK?   - WinRAR - Fast to open and extract. Silently fails (shows empty string) when attempting to read comment (beyond their defined limit).
    OK?   - WinZip - Fast to open and extract. Shows failure window when attempting to read comment (too long for them).
    OK    - Windows Explorer - Fast to open and extract. Cannot show comment to user.
    OK    - unzip (Ubuntu) - Moderately fast to opend anextract. Prints comment automatically.
    OK    - Python 3.7.4 - Moderately fast to open and extract. Has public API to access comment string.
    OK    - Java OpenJDK 12.0.2 (2019-07-16) - Moderately fast to open and extract. Has public API to access comment string.
    OK    - corefx (without fix) - Fast to open and extract. Does not have public API to access comment string.
    OK    - corefx (with fix) - Fast to open and extract. Does not have public API to access comment string.

carlossanlop commented:
Related: We have an issue open proposing to add a public API to expose the zip archive comment: https://github.com/dotnet/corefx/pull/41007

ericstj commented:
@carlossanlop can you also test a zip that has trailing data > 64KB across this matrix?  also one that has both trailing and preceeding data > 64KB?  Just so we get an understanding of how tolerant various tools are to irrelevant data in the stream.  Let me know if you need any help constructing such test files.

ericstj commented:
Great analysis so far, by the way.  Love the detail.

carlossanlop commented:
@ericstj:
> _can you also test a zip that has trailing data > 64KB across this matrix? also one that has both trailing and preceeding data > 64KB? Just so we get an understanding of how tolerant various tools are to irrelevant data in the stream._

Below are the additional requested tests. Summary:

File 1: Added 256 trailing bytes with HxD.
File 2: The same file as 1, but then copied it into a new one so Windows automatically corrects the ending of the file. The results were strangely different.
File 3: The same process as in 1+2, but instead of adding 256 trailing bytes, I added 64KB.
File 4: Edited the file in 2, and added the extra 256 preceding bytes with HxD.


---
_Comment added via Python (max length is 65,535 ASCII chars (64KB)) and 256 extra bytes added at the end using HxD._
**LargeWithEOCDCommentOf64KB_Trailing256B.zip**
>large.cab (549 MB) - _downloaded from original issue (cabs are not a zip file, so they don't have an EOCD)_
>>_...Various files and folders..._

    OK    - 7-Zip - Fast to open and extract. Does not implement feature to show comment to user.
    OK?   - WinRAR - Fast to open and extract. Silently fails (shows empty string) when attempting to read comment (beyond their defined limit).
    OK?   - WinZip - Fast to open and extract. Shows failure window when attempting to read comment (too long for them).
    FAIL  - Windows Explorer - Error dialog: "Windows cannot open the folder. The compressed (zipped) Folder 'LargeWithEOCDCommentOf64KB_Plus256B.zip' is invalid.".
    OK?   - unzip (Ubuntu) - Fast to open and extract. Prints comment automatically, but only up to 64KB of it, the rest is ignored.
    FAIL  - Python 3.7.4 - Opens the file but quickly and silently fails to extract (no exception). Has public API to access comment string, it reports a comment length of 0.
    OK    - Java OpenJDK 12.0.2 (2019-07-16) - Fast to open and extract. Does not have public API to access comment string.
    OK    - corefx (without fix) - Fast to open and extract. Does not have public API to access comment string.
    FAIL  - corefx (with fix) - Quickly throws: System.IO.InvalidDataException : End of Central Directory record could not be found.

---
_Comment added via Python (max length is 65,535 ASCII chars (64KB)) and 256 extra bytes added at the end using HxD, then copied (or cut and pasted) the file into a new location, which causes Windows to fix the ending of the file with some additional bits._
**LargeWithEOCDCommentOf64KB_Trailing256BThenCopied.zip**
>large.cab (549 MB) - _downloaded from original issue (cabs are not a zip file, so they don't have an EOCD)_
>>_...Various files and folders..._

    OK      - 7-Zip - Fast to open and extract. Does not implement feature to show comment to user.
    FAIL    - WinRAR - Can open the file and shows the correct contents, but attempting to extract the file shows an error dialog: "The archive is corrupt".
    FAIL    - WinZip - Can open the file, but shows the contents as empty. The "unzip here" right click options fails with "No files to process".
    FAIL    - Windows Explorer - Can open the file, but shows the contents as empty.
    FAIL    - unzip (Ubuntu) - Shows error: "start of central directory not found; zipfile corrupt.".
    FAIL    - Python 3.7.4 - Opens the file but quickly and silently fails to extract (no exception). Has public API to access comment string, it reports a comment length of 0.
    OK      - Java OpenJDK 12.0.2 (2019-07-16) - Fast to open and extract. Does not have public API to access comment string.
    FAIL    - corefx (without fix) - Opens the file but quickly and silently fails to extract (no exception). Does not have public API to access comment string.
    FAIL    - corefx (with fix) - Opens the file but quickly and silently fails to extract (no exception). Does not have public API to access comment string.


---
_Comment added via Python (max length is 65,535 ASCII chars (64KB)) and 64KB added at the end using HxD, then copied (or cut and pasted) the file into a new location, which causes Windows to fix the ending of the file with some additional bits._
**LargeWithEOCDCommentOf64KB_Trailing64KBThenCopied.zip**
>large.cab (549 MB) - _downloaded from original issue (cabs are not a zip file, so they don't have an EOCD)_
>>_...Various files and folders..._

    OK      - 7-Zip - Fast to open and extract. Does not implement feature to show comment to user.
    FAIL    - WinRAR - Can open the file and shows the correct contents, but attempting to extract the file shows an error dialog: "The archive is corrupt".
    FAIL    - WinZip - Can open the file, but shows the contents as empty. The "unzip here" right click options fails with "No files to process".
    FAIL    - Windows Explorer - Can open the file, but shows the contents as empty.
    FAIL    - unzip (Ubuntu) - Shows error: "start of central directory not found; zipfile corrupt.".
    FAIL    - Python 3.7.4 - Opens the file but quickly and silently fails to extract (no exception). Has public API to access comment string, it reports a comment length of 0.
    OK      - Java OpenJDK 12.0.2 (2019-07-16) - Moderately fast to open and extract. Does not have public API to access comment string.
    FAIL    - corefx (without fix) - Opens the file but quickly and silently fails to extract (no exception). Does not have public API to access comment string.
    FAIL    - corefx (with fix) - Opens the file but quickly and silently fails to extract (no exception). Does not have public API to access comment string.


---
_Comment added via Python (max length is 65,535 ASCII chars (64KB)) and 256 extra bytes added both at the beginning and at the end using HxD._
**LargeWithEOCDCommentOf64KB_TrailingAndPreceding256B.zip** - 
>large.cab (549 MB) - _downloaded from original issue (cabs are not a zip file, so they don't have an EOCD)_
>>_...Various files and folders..._

    FAIL    - 7-Zip - Fails to open the file.
    OK?     - WinRAR - Can open the file, but strangely opens large.cab directly, and allows extracting. The option to see the comment disappears from the menu bar.
    FAIL    - WinZip - Can open the file but shows the contents as empty. The "unzip here" right click options fails with "No files to process".
    FAIL    - Windows Explorer - Can open the file, but shows the contents as empty.
    FAIL    - unzip (Ubuntu) - Shows error: "start of central directory not found; zipfile corrupt.".
    FAIL    - Python 3.7.4 - Opens the file but quickly and silently fails to extract (no exception). Has public API to access comment string, it reports a comment length of 0.
    FAIL    - Java OpenJDK 12.0.2 (2019-07-16) - Opens the file but quickly and silently fails to extract (no exception). Does not have public API to access comment string.
    FAIL    - corefx (without fix) - Opens the file but quickly and silently fails to extract (no exception). Does not have public API to access comment string.
    FAIL    - corefx (with fix) - Opens the file but quickly and silently fails to extract (no exception). Does not have public API to access comment string.

danmosemsft commented:
Great testing @carlossanlop!

carlossanlop commented:
Had a conversation with Eric, and I executed my tests again, but this time avoiding Explorer from adding the trailing data when copy pasting the zip file:

---

**Large.zip (575,758,336‬ bytes)** + 64KB comment (previously added with Python)
>large.cab (549 MB) - _downloaded from original issue (cabs are not a zip file, so they don't have an EOCD)_
>>_...Various files and folders..._

    OK  7-zip - opens and extracts fast and successfully. Can show a part of the comment string via File->Properties. If copied and pasted somewhere else, the full comment can be read.
    OK? WinRAR - opens fast, but extracts slow. Silently fails (shows empty string) when attempting to read comment (beyond their defined limit).
    OK? WinZip - opens fast, but extracts slow. Shows failure window when attempting to read comment (too long for them).
    OK? Windows Explorer - opens fast, but extracts slow. Shows snippet of comment by hovering on the file.
    OK  Python - opens and extracts fast and successfully. Has public API to read and modify comment string, reads a comment length of 65535.
    OK  Java - opens and extracts fast and successfully. Does not have public API to access comment string.
    OK  corefx (without fix) - opens and extracts fast and successfully. Does not have public API to access comment string.
    OK  corefx (with fix) - opens and extracts fast and successfully. Does not have public API to access comment string.

---
**Large.zip (575,823,871‬ bytes)** + 64KB comment + 64KB trailing garbage (added with HxD)
>large.cab (549 MB) - _downloaded from original issue (cabs are not a zip file, so they don't have an EOCD)_
>>_...Various files and folders..._

    OK      7-zip - opens and extracts fast and successfully. Can show a part of the comment string via File->Properties. If copied and pasted somewhere else, the full comment can be read, and it successfully ignores the trailing text.
    FAIL    WinRAR - opens successfully and shows the contents, but fails to extract by showing the dialog "The archive is corrupt". Silently fails (shows empty string) when attempting to read comment (beyond their defined limit).
    FAIL    WinZip - unable to open. Shows the dialog "Cannot open file. It does not appear to be a valid archive. If you downloaded this file, try downloading it again.".
    FAIL    Windows Explorer - unable to open. Shows the dialog: "Windows cannot open the folder. The compressed (zipped) folder is invalid.". Hovering the file does not show the comment anymore.
    FAIL    Python - Opens successfully but fails to read file list. Has public API to read and modify comment string, but detects length of 0.
    OK      Java - Opens and extracts fast and successfully. Does not have public API to read comment.
    FAIL    corefx (without fix) - Opens successfully but fails to read file list. Does not have public API to read comment.
    FAIL    corefx (with fix) - Opens successfully but fails to read file list. Does not have public API to read comment.

**Large.zip (575,889,406‬ bytes)** + 64KB comment + 64KB trailing garbage + 64KB of preceding garbage (added with HxD)
>large.cab (549 MB) - _downloaded from original issue (cabs are not a zip file, so they don't have an EOCD)_
>>_...Various files and folders..._

    FAIL    7-zip - Unable to open, shows dialog "Can not open file large.zip as archive.".
    OK?     WinRAR - Opens the zip file, but does not show large.cab directly. Instead, shows the contents of large.cab (a bunch of package*.cab files), which can be successfully extracted. Does not show the comment button.
    FAIL    WinZip - Unable to open, but does not show any failure dialogs like in the previous case.
    FAIL    Windows Explorer - Unable to open but does not show any failure dialogs like in the previous case.
    FAIL    Python - Opens successfully but fails to read file list. Has public API to read and modify comment string, but detects length of 0.
    FAIL    Java - Opens file but fails to read file list. Does not throw. Does not have public API to read comment.
    FAIL    corefx (without fix) - Unable to open. Throws "Unhandled exception. System.IO.InvalidDataException: Number of entries expected in End Of Central Directory does not correspond to number of entries in Central Directory."
    FAIL    corefx (with fix) - Unable to open. Throws "System.IO.InvalidDataException : End of Central Directory record could not be found."

ericstj commented:
This food.  One question:

> Large.zip (575,823,871‬ bytes) + 64KB comment + 64KB trailing garbage (added with HxD)
> FAIL    corefx (without fix) - Opens successfully but fails to read file list. Does not have public API to read comment.
> FAIL    corefx (with fix) - Opens successfully but fails to read file list. Does not have public API to read comment.

I thought corefx (without fix) would succeed here?  Shouldn't this case go down the same codepath as LargeWithEOCDCommentOf64KB_Trailing256B.zip?  If it fails in both cases then that means this fix is even less breaking then we originally were thinking.

carlossanlop commented:
> _then that means this fix is even less breaking then we originally were thinking._

That seems to be the case, @ericstj. Summarized, this is what is happening when reading a zip file that has an EOCD comment + trailing garbage + preceding garbage:

Before the fix, we would:
1. Successfully create a ZipArchive object: We would read the whole file from the end to search for the EOCD marker and we would eventually find it, even if we had a lot of unexpected trailing garbage.
2. Fail to enumerate the ZipAchiveEntries: this is the part where we would attempt to read the CD, which happens from the beginning of the file, and our code expects the beginning of the file to be a valid zip structure, which in this case it's not, because of the unexpected preceding garbage. This is why we throw a confusing message that doesn't seem to explain the real cause of failure: ` System.IO.InvalidDataException: Number of entries expected in End Of Central Directory does not correspond to number of entries in Central Directory`.

After the fix we now:
1. Fail to create a ZipArchive object: We attempt to find the EOCD marker from the end of the file, up to the max expected number of bytes that can exist after the marker. Since we read garbage, we throw, and the message is now telling the user the real cause of the failure: `System.IO.InvalidDataException : End of Central Directory record could not be found`.

carlossanlop commented:
I'm working on the unit tests. I'll update the PR soon.

carlossanlop commented:
@ericstj The unit test is passing locally, but it's going to be failing until I get the dependency zip file merged and this PR fast-forwarded to the bits that can consume it.

ericstj commented:
> which in this case it's not, because of the unexpected **preceding** garbage.

I was actually looking at the case of only **trailing** garbage.  Did you figure out why it failed with only trailing garbage?  Again, I'm not saying its a problem, it's just that makes this change less breaking and it would be good to understand the changes to failure modes.

carlossanlop commented:
> I was actually looking at the case of only **trailing** garbage. Did you figure out why it failed with only trailing garbage? Again, I'm not saying its a problem, it's just that makes this change less breaking and it would be good to understand the changes to failure modes.

I'll look into this today.

I fast-forwarded my branch to latest, but it's unable to find the package, which I already merged into corefx-testdata. @ericstj would you mind helping me figure out what am I missing? This is the CI error message:

```
src\System.IO.Compression.ZipFile\tests\System.IO.Compression.ZipFile.Tests.csproj(0,0):
error NU1102: (NETCORE_ENGINEERING_TELEMETRY=Restore) Unable to find package System.IO.Compression.TestData with version (>= 1.0.15)
```

ericstj commented:
IIRC we don't have a build of test-data.  You need to build it yourself and push the package.  Is that still the case @ViktorHofer @safern?

ViktorHofer commented:
Instructions for that can be found here: https://github.com/dotnet/corefx-testdata/blob/master/README.md

carlossanlop commented:
Right, I forgot about that step 😄 thanks @ViktorHofer and @ericstj 
I uploaded the nupkg. I'm closing and reopening the PR so the CI gets triggered again.

ViktorHofer commented:
Why are you closing and re-opening this PR over and over again?

ericstj commented:
Looks to me like lots of the compression tests are failing again.  https://dev.azure.com/dnceng/public/_build/results?buildId=367365&view=ms.vss-test-web.build-test-results-tab&runId=11123408&resultId=152643&paneView=debug

This smells like the line-endings issue we hit before with these.

riverar commented:
Please figure out how to run your pipelines/tests without closing/re-opening the issue. This sends a lot of email out! Thanks! 👋

carlossanlop commented:
> Why are you closing and re-opening this PR over and over again?

Because the "Re-run" button is not always available. I realized that closing and reopening triggers CI.

safern commented:
> Because the "Re-run" button is not always available. I realized that closing and reopening triggers CI.

You can also add a comment "/azp run corefx-ci"

carlossanlop commented:
>> _Large.zip (575,823,871‬ bytes) + 64KB comment + 64KB trailing garbage (added with HxD)_
>> _FAIL corefx (without fix) - Opens successfully but fails to read file list. Does not have public API to read comment._
>> _FAIL corefx (with fix) - Opens successfully but fails to read file list. Does not have public API to read comment._

> _I thought corefx (without fix) would succeed here? Shouldn't this case go down the same codepath as LargeWithEOCDCommentOf64KB_Trailing256B.zip? If it fails in both cases then that means this fix is even less breaking then we originally were thinking._

Ok so I looked into this, @ericstj. Not sure why the code failed back when I ran this test. I did everything again, carefully, and the test actually passed, as you were expecting.

I edited the test result to:

>> OK corefx (without fix) - Opens and extracts successfully. Does not have public API to read comment.

carlossanlop commented:
The CI passed, @ericstj / @ViktorHofer / @ahsonkhan , and I answered Eric's question. Would you mind taking a look again and sign off if you feel it's ready?

ericstj commented:
I see, so we will potentially see some folks regressed by this if they happen to be extracting a zip file with garbage at the end.  Let's accept that breaking change and doc it for now.  If we hear from a lot of folks that they are broken we can revisit.

ViktorHofer commented:
This is a reminder to re-run CI if changes in the same area happened recently or if the CI hasn't run on the PR for some time. See https://github.com/dotnet/corefx/blob/master/Documentation/project-docs/pull-request-policy.md for more info.
